# Tank detection YoloV8 model train

This notebook will train a [Yolov8](https://github.com/ultralytics/ultralytics) model for tank detection using publicly available annotated images of tanks.

As the notebook will run the training with `PyTorch`, it is recommended to have GPUs available. If running in Google Colab, go to Edit > Notebook settings and select GPU hardware acceleration.

The first step of the notebook will use [fiftyone](https://github.com/voxel51/fiftyone), to build a dataset of tank images and bounding box annotations for training the computer vision model. The tutorial notebook on [Fine-tuning YOLOv8 models for custom use cases](https://github.com/voxel51/fiftyone/blob/v0.21.0/docs/source/tutorials/yolov8.ipynb) is a usefull introduction on how to use `fiftyone`.

## Setup

In [1]:
%load_ext autoreload
%autoreload 2

To start, check GPU support.

In [2]:
import torch

print(f"Setup complete. Using torch {torch.__version__} ({torch.cuda.get_device_properties(0).name if torch.cuda.is_available() else 'CPU'})")

Setup complete. Using torch 2.3.0+cu121 (NVIDIA RTX A6000)


We also setup logs.

In [3]:
import logging

logging.basicConfig(level=logging.INFO)

## Download images from ImageNet

The first dataset we'll use is ImageNet21k. The ImageNet21k dataset is available at [https://image-net.org/download-images.php](https://image-net.org/download-images.php). You need to register and be granted access to download the images. We use the Winter 21 version since it gives the option of downloading the images for a single synset: https://image-net.org/data/winter21_whole/SYNSET_ID.tar, e.g., https://image-net.org/data/winter21_whole/n02352591.tar. The processed version of ImageNet21k is available here : https://github.com/Alibaba-MIIL/ImageNet21K. The class ids and names are available here https://github.com/google-research/big_transfer/issues/7#issuecomment-640048775.

We'll begin by downloading the class names that are in ImageNet21k and look for relevant classes that we can use.

In [4]:
from pathlib import Path

imagenet_dir = Path() / "imagenet"

In [5]:
from adomvi.datasets.imagenet import download_class_names, find_class_by_text

classes = download_class_names(imagenet_dir)
find_class_by_text(classes, "military")

INFO:root:File imagenet/imagenet21k_wordnet_ids.txt already exists. Skipping download.
INFO:root:File imagenet/imagenet21k_wordnet_lemmas.txt already exists. Skipping download.


{'n03762982': 'military_hospital',
 'n03763727': 'military_quarters',
 'n03763968': 'military_uniform',
 'n03764276': 'military_vehicle',
 'n04552348': 'warplane, military_plane',
 'n08249459': 'concert_band, military_band',
 'n09809538': 'army_engineer, military_engineer',
 'n09943239': 'commissioned_military_officer',
 'n10316360': 'military_attache',
 'n10316527': 'military_chaplain, padre, Holy_Joe, sky_pilot',
 'n10316862': 'military_leader',
 'n10317007': 'military_officer, officer',
 'n10317500': 'military_policeman, MP',
 'n10512372': 'recruit, military_recruit',
 'n10582746': 'serviceman, military_man, man, military_personnel',
 'n10759331': 'volunteer, military_volunteer, voluntary'}

We can now download images and annotations for the relevant classes. The `download_imagenet_detections` function will download the images and annotations for the given class ids **if the annotations exist** (not all classes have been annotated).

In [6]:
from adomvi.datasets.imagenet import download_imagenet_detections

class_ids = ["n02740300", "n04389033", "n02740533", "n04464852", "n03764276"]
download_imagenet_detections(class_ids, imagenet_dir)

INFO:root:File imagenet/bboxes_annotations.tar.gz already exists. Skipping download.
INFO:root:There are not annotations for class n02740300.
INFO:root:Annotations directory imagenet/labels/n04389033 already exists. Skipping extract.
INFO:root:There are not annotations for class n02740533.
INFO:root:There are not annotations for class n04464852.
INFO:root:There are not annotations for class n03764276.
INFO:root:Deleting annotations dir.


The data we just downloaded into the `imagenet` directory is not all clean: there are annotations which have no corresponding image. We need to remove those labels, otherwise this causes errors when importing the data into fiftyone.

In [7]:
from adomvi.datasets.imagenet import cleanup_labels_without_images

cleanup_labels_without_images(imagenet_dir)

INFO:root:Deleting 0 labels without images


We can now create a new dataset with `fiftyone`. Fiftyone allows us to manage images annotated with bounding boxes and labels, to merge datasets from different sources, and to split the datasets and prepare them for processing.

In [8]:
import fiftyone as fo

# Create the dataset
dataset = fo.Dataset.from_dir(
    dataset_dir=imagenet_dir,
    dataset_type=fo.types.VOCDetectionDataset,
)

dataset.name = "military-vehicles"

dataset.map_labels(
    "ground_truth",
    {"n04389033":"tank"}
).save()

 100% |█████████████████| 378/378 [458.6ms elapsed, 0s remaining, 827.5 samples/s]      


INFO:eta.core.utils: 100% |█████████████████| 378/378 [458.6ms elapsed, 0s remaining, 827.5 samples/s]      


Once our dataset is created, we can launch a session to display the dataset and view the annotated images

In [9]:
session = fo.launch_app(dataset, auto=False)

Session launched. Run `session.show()` to open the App in a cell output.


INFO:fiftyone.core.session.session:Session launched. Run `session.show()` to open the App in a cell output.


In [10]:
session.open_tab()

<IPython.core.display.Javascript object>

### Add OpenImage samples

The ImageNet dataset only contained 378 annotated images of tanks, so we'll look into other available datasets to improve training of the model. We’ll load [Open Images](https://storage.googleapis.com/openimages/web/index.html) samples with `Tank` detection labels, passing in `only_matching=True` to only load the `Tank` labels. We then map these labels by changing `Tank` into `tank`.

In [11]:
import fiftyone.zoo as foz

oi_samples = foz.load_zoo_dataset(
    "open-images-v7",
    classes = ["Tank"],
    only_matching=True,
    label_types="detections"
).map_labels(
    "ground_truth",
    {"Tank":"tank"}
)

INFO:fiftyone.zoo.datasets:Downloading split 'train' to '/home/jrenault/fiftyone/open-images-v7/train' if necessary


Necessary images already downloaded


INFO:fiftyone.utils.openimages:Necessary images already downloaded


Existing download of split 'train' is sufficient


INFO:fiftyone.zoo.datasets:Existing download of split 'train' is sufficient


INFO:fiftyone.zoo.datasets:Downloading split 'test' to '/home/jrenault/fiftyone/open-images-v7/test' if necessary


Necessary images already downloaded


INFO:fiftyone.utils.openimages:Necessary images already downloaded


Existing download of split 'test' is sufficient


INFO:fiftyone.zoo.datasets:Existing download of split 'test' is sufficient


INFO:fiftyone.zoo.datasets:Downloading split 'validation' to '/home/jrenault/fiftyone/open-images-v7/validation' if necessary


Necessary images already downloaded


INFO:fiftyone.utils.openimages:Necessary images already downloaded


Existing download of split 'validation' is sufficient


INFO:fiftyone.zoo.datasets:Existing download of split 'validation' is sufficient


Loading 'open-images-v7' split 'train'


INFO:fiftyone.zoo.datasets:Loading 'open-images-v7' split 'train'


 100% |███████████████| 1062/1062 [1.7s elapsed, 0s remaining, 636.0 samples/s]         


INFO:eta.core.utils: 100% |███████████████| 1062/1062 [1.7s elapsed, 0s remaining, 636.0 samples/s]         


Loading 'open-images-v7' split 'test'


INFO:fiftyone.zoo.datasets:Loading 'open-images-v7' split 'test'


 100% |█████████████████| 134/134 [212.8ms elapsed, 0s remaining, 629.6 samples/s]      


INFO:eta.core.utils: 100% |█████████████████| 134/134 [212.8ms elapsed, 0s remaining, 629.6 samples/s]      


Loading 'open-images-v7' split 'validation'


INFO:fiftyone.zoo.datasets:Loading 'open-images-v7' split 'validation'


 100% |███████████████████| 50/50 [85.3ms elapsed, 0s remaining, 586.3 samples/s]     


INFO:eta.core.utils: 100% |███████████████████| 50/50 [85.3ms elapsed, 0s remaining, 586.3 samples/s]     


Dataset 'open-images-v7' created


INFO:fiftyone.zoo.datasets:Dataset 'open-images-v7' created


In [12]:
dataset.merge_samples(oi_samples)

## Fine-tune a YOLOv8 detection model

Now that our dataset is created, we'll export it into a format supported by YOLOv8 to train our model.

We first remove tags from the dataset, and split it into a train, val and test sets.

In [13]:
import fiftyone.utils.random as four

## delete existing tags to start fresh
dataset.untag_samples(dataset.distinct("tags"))

## split into train, test and val
four.random_split(dataset, {"train": 0.8, "val": 0.1, "test": 0.1})

Once our dataset is split, we can export it to a specific directory.

In [17]:
from pathlib import Path
from adomvi.yolo.utils import export_yolo_data

export_dir = Path() / "vehicles"
export_yolo_data(dataset, export_dir, ["tank"], split = ["train", "val", "test"], overwrite=True)

 100% |███████████████| 1299/1299 [1.3s elapsed, 0s remaining, 979.9 samples/s]        


INFO:eta.core.utils: 100% |███████████████| 1299/1299 [1.3s elapsed, 0s remaining, 979.9 samples/s]        


Directory 'vehicles' already exists; export will be merged with existing files


 100% |█████████████████| 163/163 [166.7ms elapsed, 0s remaining, 977.8 samples/s]     


INFO:eta.core.utils: 100% |█████████████████| 163/163 [166.7ms elapsed, 0s remaining, 977.8 samples/s]     


Directory 'vehicles' already exists; export will be merged with existing files


 100% |█████████████████| 162/162 [189.8ms elapsed, 0s remaining, 853.4 samples/s]     


INFO:eta.core.utils: 100% |█████████████████| 162/162 [189.8ms elapsed, 0s remaining, 853.4 samples/s]     


Finally, we can train a YoloV8 base model on our exported dataset. We use the `yolov8n.pt` (nano) model, which is the smallest, but larger models are available from ultralytics. Simply replace `yolov8n.pt` with the [right model](https://docs.ultralytics.com/models/yolov8/#key-features).

In [19]:
from adomvi.yolo.yolo import train

train("yolov8n.pt", data=export_dir / "dataset.yaml", batch=64, device=[0, 1])

Ultralytics YOLOv8.2.28 🚀 Python-3.10.14 torch-2.3.0+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
                                                            CUDA:1 (NVIDIA RTX A6000, 48655MiB)
engine/trainer: task=detect, mode=train, model=yolov8n.pt, data=vehicles/dataset.yaml, epochs=60, time=None, patience=100, batch=64, imgsz=640, save=True, save_period=-1, cache=False, device=[0, 1], workers=8, project=None, name=train, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_fr

/home/jrenault/mambaforge/envs/adomvi2/lib/python3.10/site-packages/torch/nn/modules/conv.py:456: UserWarning: Plan failed with a cudnnException: CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR: cudnnFinalize Descriptor Failed cudnn_status: CUDNN_STATUS_NOT_SUPPORTED (Triggered internally at ../aten/src/ATen/native/cudnn/Conv_v8.cpp:919.)
  return F.conv2d(input, weight, bias, self.stride,


AMP: checks passed ✅


train: Scanning /home/jrenault/workspace/adomvi2/notebooks/vehicles/labels/train.cache... 1299 images, 0 backgrounds, 0 corrupt: 100%|██████████| 1299/1299 [00:00<?, ?it/s]
val: Scanning /home/jrenault/workspace/adomvi2/notebooks/vehicles/labels/val.cache... 163 images, 0 backgrounds, 0 corrupt: 100%|██████████| 163/163 [00:00<?, ?it/s]


Plotting labels to runs/detect/train/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000714, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 16 dataloader workers
Logging results to runs/detect/train
Starting training for 60 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


  0%|          | 0/21 [00:00<?, ?it/s]/home/jrenault/mambaforge/envs/adomvi2/lib/python3.10/site-packages/torch/nn/modules/conv.py:456: UserWarning: Plan failed with a cudnnException: CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR: cudnnFinalize Descriptor Failed cudnn_status: CUDNN_STATUS_NOT_SUPPORTED (Triggered internally at ../aten/src/ATen/native/cudnn/Conv_v8.cpp:919.)
  return F.conv2d(input, weight, bias, self.stride,
/home/jrenault/mambaforge/envs/adomvi2/lib/python3.10/site-packages/torch/nn/modules/conv.py:456: UserWarning: Plan failed with a cudnnException: CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR: cudnnFinalize Descriptor Failed cudnn_status: CUDNN_STATUS_NOT_SUPPORTED (Triggered internally at ../aten/src/ATen/native/cudnn/Conv_v8.cpp:919.)
  return F.conv2d(input, weight, bias, self.stride,
       1/60      4.33G     0.9737      2.306       1.37         30        640: 100%|██████████| 21/21 [00:05<00:00,  3.59it/s]
                 Class     Images  Instances      Box(P          

                   all        163        245     0.0906      0.873       0.48      0.262

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/60       4.3G      0.924      1.562      1.357         26        640: 100%|██████████| 21/21 [00:03<00:00,  6.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.82it/s]


                   all        163        245      0.944      0.415      0.697      0.438

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/60      4.35G     0.9284      1.423      1.337         43        640: 100%|██████████| 21/21 [00:03<00:00,  6.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.95it/s]


                   all        163        245      0.772       0.58      0.674      0.417

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/60       4.4G     0.9511      1.361      1.356         33        640: 100%|██████████| 21/21 [00:03<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.39it/s]


                   all        163        245      0.715      0.542      0.606      0.292

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/60      4.34G     0.9505      1.304      1.348         32        640: 100%|██████████| 21/21 [00:03<00:00,  6.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.05it/s]


                   all        163        245      0.657      0.604      0.625      0.361

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/60       4.3G     0.9715      1.242      1.354         37        640: 100%|██████████| 21/21 [00:03<00:00,  6.07it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.65it/s]


                   all        163        245      0.716      0.624      0.675      0.397

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/60      4.34G     0.9598      1.201       1.36         37        640: 100%|██████████| 21/21 [00:03<00:00,  6.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.70it/s]


                   all        163        245      0.612      0.539      0.573      0.285

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/60       4.3G     0.9501      1.135      1.331         37        640: 100%|██████████| 21/21 [00:03<00:00,  6.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.84it/s]


                   all        163        245      0.753      0.645      0.697       0.42

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/60      4.34G     0.9774      1.113      1.357         34        640: 100%|██████████| 21/21 [00:03<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.93it/s]


                   all        163        245      0.755      0.579      0.677        0.4

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/60      4.34G     0.9283       1.04       1.32         37        640: 100%|██████████| 21/21 [00:03<00:00,  6.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.65it/s]


                   all        163        245      0.702      0.664      0.712      0.422

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/60      4.34G     0.9155      1.017      1.307         36        640: 100%|██████████| 21/21 [00:03<00:00,  5.96it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.41it/s]


                   all        163        245      0.722      0.679      0.722      0.413

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/60       4.3G     0.9083     0.9753      1.307         42        640: 100%|██████████| 21/21 [00:03<00:00,  5.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.95it/s]


                   all        163        245      0.884      0.657      0.762      0.463

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/60      4.34G     0.8568      0.903      1.257         27        640: 100%|██████████| 21/21 [00:03<00:00,  5.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.28it/s]


                   all        163        245      0.792      0.718      0.778      0.489

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/60       4.3G     0.8995     0.9302      1.296         33        640: 100%|██████████| 21/21 [00:03<00:00,  6.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.16it/s]


                   all        163        245      0.768      0.731      0.757      0.456

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/60      4.35G     0.8592     0.9058      1.282         22        640: 100%|██████████| 21/21 [00:03<00:00,  6.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.18it/s]


                   all        163        245       0.84      0.667      0.764      0.475

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/60      4.34G     0.8632     0.8712      1.282         30        640: 100%|██████████| 21/21 [00:03<00:00,  6.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.32it/s]


                   all        163        245      0.766      0.653      0.722      0.461

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/60       4.3G     0.8511     0.8583      1.247         40        640: 100%|██████████| 21/21 [00:03<00:00,  6.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.89it/s]


                   all        163        245      0.835      0.722      0.797      0.542

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/60      4.29G     0.8379     0.8717      1.246         32        640: 100%|██████████| 21/21 [00:03<00:00,  5.98it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.59it/s]


                   all        163        245      0.811      0.731      0.812      0.529

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/60      4.34G     0.8002     0.8066       1.23         24        640: 100%|██████████| 21/21 [00:03<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.50it/s]


                   all        163        245      0.789      0.739      0.805      0.528

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/60      4.29G     0.8032     0.8021      1.229         26        640: 100%|██████████| 21/21 [00:03<00:00,  6.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.95it/s]


                   all        163        245      0.836      0.731      0.793      0.473

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      21/60      4.29G     0.7992      0.754      1.214         28        640: 100%|██████████| 21/21 [00:03<00:00,  6.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.65it/s]


                   all        163        245      0.826      0.694      0.755      0.467

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      22/60      4.34G     0.7834     0.7734      1.212         43        640: 100%|██████████| 21/21 [00:03<00:00,  6.15it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.02it/s]


                   all        163        245      0.805      0.763      0.813      0.531

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      23/60      4.35G     0.7869     0.7361      1.221         33        640: 100%|██████████| 21/21 [00:03<00:00,  6.07it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.36it/s]


                   all        163        245      0.862      0.751      0.826      0.551

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      24/60      4.34G     0.7922     0.7488      1.216         38        640: 100%|██████████| 21/21 [00:03<00:00,  5.96it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.71it/s]


                   all        163        245      0.809      0.726      0.794      0.504

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      25/60      4.33G     0.7597     0.7287      1.182         36        640: 100%|██████████| 21/21 [00:03<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.90it/s]


                   all        163        245      0.846      0.716      0.807      0.517

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      26/60      4.34G     0.7751      0.726      1.202         44        640: 100%|██████████| 21/21 [00:03<00:00,  6.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.88it/s]


                   all        163        245      0.776      0.784      0.813      0.539

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      27/60      4.33G      0.754     0.7149      1.195         35        640: 100%|██████████| 21/21 [00:03<00:00,  6.17it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.94it/s]


                   all        163        245      0.822      0.763      0.816      0.534

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      28/60      4.34G     0.7481     0.6883      1.184         24        640: 100%|██████████| 21/21 [00:03<00:00,  5.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  3.94it/s]


                   all        163        245      0.851       0.78      0.832      0.559

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      29/60      4.34G     0.7233     0.6737      1.171         35        640: 100%|██████████| 21/21 [00:03<00:00,  5.57it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.22it/s]


                   all        163        245      0.808      0.753      0.781      0.524

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      30/60      4.34G     0.7155     0.6636      1.176         37        640: 100%|██████████| 21/21 [00:03<00:00,  5.80it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.80it/s]


                   all        163        245      0.805      0.771      0.795      0.523

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      31/60      4.29G     0.7123       0.66      1.165         33        640: 100%|██████████| 21/21 [00:03<00:00,  5.69it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.24it/s]


                   all        163        245      0.916      0.727       0.83      0.557

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      32/60      4.29G     0.6986     0.6265      1.146         27        640: 100%|██████████| 21/21 [00:03<00:00,  5.63it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.15it/s]


                   all        163        245      0.837      0.763      0.834      0.567

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      33/60      4.34G      0.691      0.597      1.145         26        640: 100%|██████████| 21/21 [00:03<00:00,  5.63it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.26it/s]


                   all        163        245      0.916      0.715      0.847      0.578

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      34/60      4.35G     0.7043     0.6216      1.161         31        640: 100%|██████████| 21/21 [00:03<00:00,  5.68it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  3.61it/s]


                   all        163        245      0.784        0.8       0.81      0.569

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      35/60      4.29G     0.6978     0.6269      1.158         38        640: 100%|██████████| 21/21 [00:03<00:00,  5.71it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.34it/s]


                   all        163        245      0.857      0.751      0.832      0.579

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      36/60      4.34G      0.678     0.6059      1.136         29        640: 100%|██████████| 21/21 [00:03<00:00,  5.57it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.55it/s]


                   all        163        245      0.822      0.752      0.804      0.553

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      37/60      4.33G     0.6669     0.5964      1.134         23        640: 100%|██████████| 21/21 [00:03<00:00,  5.69it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.07it/s]


                   all        163        245      0.834      0.743      0.817       0.56

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      38/60      4.34G      0.659     0.5652      1.144         34        640: 100%|██████████| 21/21 [00:03<00:00,  5.60it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.42it/s]


                   all        163        245      0.821      0.816      0.859      0.603

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      39/60      4.29G     0.6479     0.5704      1.126         40        640: 100%|██████████| 21/21 [00:03<00:00,  5.51it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.57it/s]


                   all        163        245      0.874      0.755      0.839      0.587

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      40/60      4.33G     0.6414      0.581      1.106         35        640: 100%|██████████| 21/21 [00:03<00:00,  5.79it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.43it/s]


                   all        163        245      0.886      0.751      0.835       0.58

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      41/60      4.35G     0.6352     0.5652      1.111         22        640: 100%|██████████| 21/21 [00:03<00:00,  5.65it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.63it/s]


                   all        163        245      0.884      0.747      0.848       0.59

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      42/60      4.29G     0.6205     0.5371      1.107         31        640: 100%|██████████| 21/21 [00:03<00:00,  5.65it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.46it/s]


                   all        163        245      0.897      0.747      0.848      0.598

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      43/60      4.33G     0.6315     0.5305      1.109         29        640: 100%|██████████| 21/21 [00:03<00:00,  5.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.28it/s]


                   all        163        245       0.82      0.792      0.832       0.59

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      44/60      4.29G     0.6315      0.531      1.117         49        640: 100%|██████████| 21/21 [00:03<00:00,  5.58it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.17it/s]


                   all        163        245      0.865      0.767      0.832      0.589

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      45/60      4.29G      0.613      0.517      1.097         47        640: 100%|██████████| 21/21 [00:03<00:00,  5.71it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.21it/s]


                   all        163        245      0.864       0.75      0.828      0.588

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      46/60      4.29G     0.6109     0.5131      1.099         32        640: 100%|██████████| 21/21 [00:03<00:00,  5.82it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.18it/s]


                   all        163        245       0.85      0.764      0.837      0.587

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      47/60      4.35G     0.6125     0.5324       1.11         42        640: 100%|██████████| 21/21 [00:03<00:00,  5.47it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.50it/s]


                   all        163        245      0.907      0.739      0.834      0.594

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      48/60      4.34G     0.6047     0.5345      1.087         44        640: 100%|██████████| 21/21 [00:03<00:00,  5.72it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  3.77it/s]


                   all        163        245      0.853       0.78       0.84      0.592

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      49/60      4.33G     0.5947     0.5054      1.074         28        640: 100%|██████████| 21/21 [00:03<00:00,  5.51it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.29it/s]


                   all        163        245      0.835      0.803      0.836      0.588

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      50/60      4.33G     0.5863     0.4921      1.088         30        640: 100%|██████████| 21/21 [00:03<00:00,  5.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.04it/s]


                   all        163        245      0.855      0.771      0.835      0.592
Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      51/60      4.49G     0.5218     0.4896      1.041         16        640: 100%|██████████| 21/21 [00:05<00:00,  3.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.52it/s]


                   all        163        245      0.866      0.771      0.828       0.59

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      52/60       4.3G      0.492     0.4033      1.009         16        640: 100%|██████████| 21/21 [00:03<00:00,  5.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.75it/s]


                   all        163        245      0.844      0.771      0.828      0.599

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      53/60       4.3G     0.4887     0.3918      1.005         11        640: 100%|██████████| 21/21 [00:03<00:00,  5.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  3.85it/s]


                   all        163        245       0.84      0.788      0.836      0.599

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      54/60      4.29G     0.4808      0.365      1.004         13        640: 100%|██████████| 21/21 [00:03<00:00,  5.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.59it/s]


                   all        163        245      0.831      0.782      0.828      0.589

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      55/60       4.3G     0.4629     0.3694     0.9784         14        640: 100%|██████████| 21/21 [00:03<00:00,  5.86it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.63it/s]


                   all        163        245      0.839        0.8      0.828      0.591

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      56/60      4.29G     0.4789     0.3745     0.9933         20        640: 100%|██████████| 21/21 [00:03<00:00,  5.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.41it/s]


                   all        163        245      0.836      0.811      0.842      0.604

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      57/60       4.3G     0.4471     0.3624     0.9829         16        640: 100%|██████████| 21/21 [00:03<00:00,  5.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.00it/s]


                   all        163        245      0.883      0.771      0.846      0.613

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      58/60       4.3G     0.4367      0.344     0.9727         18        640: 100%|██████████| 21/21 [00:03<00:00,  5.68it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  5.86it/s]


                   all        163        245        0.9      0.767       0.85      0.618

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      59/60       4.3G     0.4265     0.3488     0.9765         10        640: 100%|██████████| 21/21 [00:03<00:00,  5.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  4.52it/s]


                   all        163        245      0.859      0.788      0.844      0.615

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      60/60       4.3G     0.4322     0.3308     0.9603         15        640: 100%|██████████| 21/21 [00:03<00:00,  5.72it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:00<00:00,  3.90it/s]


                   all        163        245       0.91      0.751      0.844      0.613

60 epochs completed in 0.081 hours.
Optimizer stripped from runs/detect/train/weights/last.pt, 6.3MB
Optimizer stripped from runs/detect/train/weights/best.pt, 6.3MB

Validating runs/detect/train/weights/best.pt...
Ultralytics YOLOv8.2.28 🚀 Python-3.10.14 torch-2.3.0+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
                                                            CUDA:1 (NVIDIA RTX A6000, 48655MiB)
Model summary (fused): 168 layers, 3005843 parameters, 0 gradients, 8.1 GFLOPs


/home/jrenault/mambaforge/envs/adomvi2/lib/python3.10/site-packages/torch/nn/modules/conv.py:456: UserWarning: Plan failed with a cudnnException: CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR: cudnnFinalize Descriptor Failed cudnn_status: CUDNN_STATUS_NOT_SUPPORTED (Triggered internally at ../aten/src/ATen/native/cudnn/Conv_v8.cpp:919.)
  return F.conv2d(input, weight, bias, self.stride,
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 3/3 [00:01<00:00,  2.71it/s]


                   all        163        245      0.902      0.767       0.85      0.618
Speed: 0.1ms preprocess, 1.0ms inference, 0.0ms loss, 0.9ms postprocess per image
Results saved to runs/detect/train


## Use the fine-tuned model to predict detections

Once our model is trained, we can simply use it to predict detections on our test dataset. Simply pass the path to the best model weights and the directory containing the test images to the `predict` function.

In [20]:
from pathlib import Path
from adomvi.yolo.yolo import predict

best_model = Path() / "runs/detect/train/weights/best.pt"
results = predict(best_model, source=export_dir / "images/test")


image 1/162 /home/jrenault/workspace/adomvi2/notebooks/vehicles/images/test/0080cbc853059db1.jpg: 448x640 1 tank, 86.1ms
image 2/162 /home/jrenault/workspace/adomvi2/notebooks/vehicles/images/test/00c5be8db0e25f5e.jpg: 448x640 1 tank, 4.8ms
image 3/162 /home/jrenault/workspace/adomvi2/notebooks/vehicles/images/test/01368575b56715e7.jpg: 640x640 1 tank, 5.3ms
image 4/162 /home/jrenault/workspace/adomvi2/notebooks/vehicles/images/test/014df43a4a199fe0.jpg: 480x640 12 tanks, 88.6ms
image 5/162 /home/jrenault/workspace/adomvi2/notebooks/vehicles/images/test/016727715ecbb0ea.jpg: 384x640 1 tank, 92.6ms
image 6/162 /home/jrenault/workspace/adomvi2/notebooks/vehicles/images/test/021bb7e1cbb12d26.jpg: 384x640 1 tank, 8.2ms
image 7/162 /home/jrenault/workspace/adomvi2/notebooks/vehicles/images/test/025457c66e740c30.jpg: 448x640 1 tank, 7.4ms
image 8/162 /home/jrenault/workspace/adomvi2/notebooks/vehicles/images/test/02766a4a2a22632b.jpg: 640x448 2 tanks, 105.9ms
image 9/162 /home/jrenault/work

## Evaluate model predictions

We can first evaluate our model by loading the predictions in a Fiftyone dataset view of our test data. We select the images tagged as part of the test set in our dataset.

In [21]:
# The test split of the dataset
test_view = dataset.match_tags("test")

Then we can load the model predictions as fields of the images in our test view.

In [35]:
from pathlib import Path
from adomvi.yolo.utils import add_yolo_detections

prediction_field = "yolov8"
predictions_dir = Path() / "runs/detect/predict/labels"
add_yolo_detections(test_view, prediction_field=prediction_field, predictions_dir=predictions_dir, class_list=["tank"])

Now we can visualize these YOLOv8 model predictions on the samples in our dataset in the FiftyOne app:

In [36]:
session.open_tab()

<IPython.core.display.Javascript object>

Finally, we can evaluate our model's predictions and print the mAP.

In [37]:
detection_results = test_view.evaluate_detections(
    prediction_field, 
    eval_key="eval",
    compute_mAP=True,
    gt_field="ground_truth",
)

Evaluating detections...


INFO:fiftyone.utils.eval.detection:Evaluating detections...


 100% |█████████████████| 162/162 [569.9ms elapsed, 0s remaining, 284.3 samples/s]      


INFO:eta.core.utils: 100% |█████████████████| 162/162 [569.9ms elapsed, 0s remaining, 284.3 samples/s]      


Performing IoU sweep...


INFO:fiftyone.utils.eval.coco:Performing IoU sweep...


 100% |█████████████████| 162/162 [585.3ms elapsed, 0s remaining, 276.8 samples/s]      


INFO:eta.core.utils: 100% |█████████████████| 162/162 [585.3ms elapsed, 0s remaining, 276.8 samples/s]      


In [38]:
mAP = detection_results.mAP()
print(f"mAP = {mAP}")

mAP = 0.5826986060627386


In [39]:
detection_results.print_report()

              precision    recall  f1-score   support

        tank       0.85      0.79      0.82       260

   micro avg       0.85      0.79      0.82       260
   macro avg       0.85      0.79      0.82       260
weighted avg       0.85      0.79      0.82       260

